File: src/models/uniquetable.cpp
#include "uniqueTable.hpp"

UniqueTable& UniqueTable::getInstance() {
    static UniqueTable instance;
    return instance;
}

void UniqueTable::insert(size_t hashKey, shared_ptr<QMDDNode> node) {
    table[hashKey].push_back(node);
}

shared_ptr<QMDDNode> UniqueTable::find(size_t hashKey) const {
        auto it = table.find(hashKey);
        if (it != table.end()) {
            return it->second[0];
        }
        return nullptr;
}

void UniqueTable::printAllEntries() const {
    for (const auto& entry : table) {
        size_t key = entry.first;
        const auto& nodes = entry.second;

        cout << "Key: " << key << endl;
        cout << "Nodes: " << endl;

        for (const auto& nodePtr : nodes) {
            if (nodePtr) {
                const QMDDNode& node = *nodePtr;
                cout << "  Node with " << node.edges.size() << " rows of children." << endl;
                for (size_t i = 0; i < node.edges.size(); ++i) {
                    for (size_t j = 0; j < node.edges[i].size(); ++j) {
                        const QMDDEdge& edge = node.edges[i][j];
                        cout << "    Edge (" << i << ", " << j << "): weight = " << edge.weight << ", Key = " << edge.uniqueTableKey << ", isTerminal = " << edge.isTerminal << endl;
                    }
                }
            } else {
                cout << "  Null node" << endl;
            }
        }
        cout << endl;
    }
}

File: src/macOS/Sources/uniqueTable.swift
import Foundation
import Numerics

class UniqueTable {
    static let instance = UniqueTable()
    private var table: [UInt: [QMDDNode]] = [:]
    private let tableQueue = DispatchQueue(label: "uniqueTable.queue", attributes: .concurrent)

    private init() {}

    static func getInstance() -> UniqueTable {
        return instance
    }

    func insert(hashKey: UInt, node: QMDDNode) {
        tableQueue.async(flags: .barrier) {
            if self.table[hashKey] != nil {
                self.table[hashKey]?.append(node)
            } else {
                self.table[hashKey] = [node]
            }
        }
    }

    func find(hashKey: UInt) -> QMDDNode? {
        var result: QMDDNode?
        tableQueue.sync {
            if let nodes = self.table[hashKey], !nodes.isEmpty {
                result = nodes[0]
            }
        }
        return result
    }

    func printAllEntries() {
        tableQueue.sync {
            for (key, nodes) in self.table {
                print("Key: \(key)")
                print("Nodes:")

                for node in nodes {
                    print("  \(node)")
                }
                print()
            }
        }
    }
}